### PR TITLE
Feature / Add enum options to deserialize relevant attributes 

### PR DESCRIPTION
**Pull request by Jerry-Jinfeng-Guo**
This PR adds to the deserialization step the possibility of deserializing only relevant data (components or attributes that are not NaN or not empty). This option will be part of `IntEnum` that is defined as: 
 - `ALL`, `0`, all data
 - `RELEVANT`, `1`, all relevant data

What's included in this PR:
- [x] Remove Ellipsis and replace with the new enum
~~- [ ] Add the logic for `RELEVANT`~~
- [x] Update tests to migrate from Ellipse to new enum
~~- [ ] Update tests for `RELEVANT` logic~~

This PR is based on main but should be further tested with https://github.com/PowerGridModel/power-grid-model/pull/699

### CODE REVIEW

File: src/power_grid_model/typing.py [hunk 1]
```diff
@@ -6,12 +6,26 @@
...
 """
 
-from types import EllipsisType
+from enum import IntEnum
 
 from power_grid_model.core.dataset_definitions import ComponentType, ComponentTypeVar
 
-_ComponentAttributeMappingDict = dict[ComponentType, set[str] | list[str] | None | EllipsisType]
+
+class ComponentAttributeFilterOptions(IntEnum):
```
Comment (mgovers) [inline src/power_grid_model/typing.py:14]: we might want to add this to `src/power_grid_model/__init__.py`

File: tests/unit/test_internal_utils.py [hunk 2]
```diff
@@ -430,18 +431,34 @@ def test_convert_batch_dataset_to_batch_list_invalid_type_sparse(_mock: MagicMoc
         convert_batch_dataset_to_batch_list(update_data)
 
 
+data_filter_all = ComponentAttributeFilterOptions.ALL
+data_filter_relevant = ComponentAttributeFilterOptions.RELEVANT
```
Comment (mgovers) [inline tests/unit/test_internal_utils.py:435]: * why do you want this? if the name is too long, you can also use the `from ... import ... as ...` syntax
* it's good practice to capitalize global variables (cfr. PEP 8 (https://peps.python.org/pep-0008/#constants))

```suggestion
DATA_FILTER_ALL = ComponentAttributeFilterOptions.ALL
DATA_FILTER_RELEVANT = ComponentAttributeFilterOptions.RELEVANT
```

File: tests/unit/test_serialization.py [hunk 3]
```diff
@@ -349,7 +353,7 @@ def split_deserialized_dataset_into_individual_scenario(scenario_idx, deserializ
 def assert_not_a_value(value: np.ndarray):
     """Assert value contains only NaN-like values.
 
-    Depending on type, maps to np.nan, -127, na_IntID, ...
+    Depending on type, maps to np.nan, -127, na_IntID, ComponentAttributeFilterOptions
```
Comment (mgovers) [inline tests/unit/test_serialization.py:356]: gotta love global search/replace

```suggestion
    Depending on type, maps to np.nan, -127, na_IntID, ...
```
or
```suggestion
    Depending on type, maps to np.nan, -127, na_IntID, etc.
```

File: tests/unit/validation/test_input_validation.py [hunk 4]
```diff
@@ -277,7 +278,8 @@ def original_data() -> dict[ComponentType, np.ndarray]:
 
 @pytest.fixture
 def original_data_columnar(original_data):
-    return compatibility_convert_row_columnar_dataset(original_data, Ellipsis, DatasetType.input)
+    data_filter = ComponentAttributeFilterOptions.ALL
+    return compatibility_convert_row_columnar_dataset(original_data, data_filter, DatasetType.input)
```
Comment (mgovers) [inline tests/unit/validation/test_input_validation.py:282]: probably good to parametrize

File: tests/unit/validation/test_batch_validation.py [hunk 5]
```diff
@@ -56,7 +57,8 @@ def original_batch_data() -> dict[str, np.ndarray]:
 
 @pytest.fixture
 def original_batch_data_columnar(original_batch_data):
-    return compatibility_convert_row_columnar_dataset(original_batch_data, Ellipsis, DatasetType.update)
+    data_filter = ComponentAttributeFilterOptions.ALL
+    return compatibility_convert_row_columnar_dataset(original_batch_data, data_filter, DatasetType.update)
```
Comment (mgovers) [inline tests/unit/validation/test_batch_validation.py:61]: probably good to parametrize

File: tests/unit/validation/test_input_validation.py (hunk 4 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/unit/validation/test_input_validation.py:282]: more specific?

Comment (mgovers) [reply tests/unit/validation/test_input_validation.py:282]: i mean to have this fixture  parametrized for both `ComponentAttributeFilterOptions.ALL` and `ComponentAttributeFilterOptions.RELEVANT`

File: src/power_grid_model/_utils.py [hunk 6]
```diff
@@ -316,7 +319,7 @@ def compatibility_convert_row_columnar_dataset(
     Args:
         data (Dataset): dataset to convert
         data_filter (ComponentAttributeMapping): desired component and attribute mapping
-        dataset_type (DatasetType): type of dataset
+        dataset_type (DatasetType): type of dataset, input, update or (sym | asym | sc)_output
```
Comment (mgovers) [inline src/power_grid_model/_utils.py:322]:
```suggestion
        dataset_type (DatasetType): type of dataset (e.g.: input, update or (sym | asym | sc)_output)
```

**Review by mgovers — APPROVED** [review]
small optional nitpick; the rest LGTM

File: tests/unit/test_internal_utils.py (hunk 2 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/unit/test_internal_utils.py]: Addressed.

File: tests/unit/test_serialization.py (hunk 3 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/unit/test_serialization.py]: Addressed.

File: src/power_grid_model/_utils.py (hunk 6 above)
Comment (Jerry-Jinfeng-Guo) [reply src/power_grid_model/_utils.py]: Addressed.

File: tests/unit/validation/test_batch_validation.py (hunk 5 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/unit/validation/test_batch_validation.py:61]: Addressed.

**Review by mgovers — APPROVED** [review]
current state LGTM for greater audience; actual in-depth implementation to be done in follow-up PR

File: tests/unit/validation/test_input_validation.py (hunk 4 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/unit/validation/test_input_validation.py]: Best attempt for now pushed. Will further polish in the follow-up.

**Review by mgovers — APPROVED** [review]

**Pull request merged**